xtext.cpp

```diff
@@ -0,0 +1,2157 @@
+            // Arabic surrounded by hebrew chars would not get its letters joined
+            // if they were all shaped as a single segment. This may probably happen
+            // too with some complex LTR scripts like indic surrounded by latin.
+            // Note: libraqm and Lua library https://github.com/luapower/tr do
+            // a bit more than that by trying to make neutral paired characters part
+            // of a same script segment (_raqm_resolve_scripts(), using a stack, so
+            // probably costly and needing another pass). We don't do that for now.
+            bool script_changed = false;
+            if ( i < m_length ) {
+                hb_script_t script = hb_unicode_script(unicode_funcs, m_text[i]);
+                if ( script != HB_SCRIPT_COMMON && script != HB_SCRIPT_INHERITED && script != HB_SCRIPT_UNKNOWN ) {
+                    if ( prev_script != HB_SCRIPT_COMMON && script != prev_script ) {
+                        m_charinfo[i].flags |= CHAR_SCRIPT_CHANGE;
+                        script_changed = true;
+                        m_has_multiple_scripts = true;
+                    }
+                    prev_script = script;
+                }
+                // Note: as we have here guessed the script of what's to be measured
+                // next, we could store it in m_charinfo (an additional int32...),
+                // so we can pass it to getHbFontData() (or use it and check
+                // ourselves here), to skip fallback fonts that do not support
+                // this script - if fonts announce the scripts they support (I think
+                // I have seen that in some font tables, may be OTF only?).
+            }
+
+            // Line breaking and wrapping
+            bool line_break = false;
+            if ( i == 0 ) {
+                lb_init_break_context(&lbCtx, m_text[i], m_lang ? m_lang : default_lang);
+            }
+            else {
+                // When at end of m_text, add a letter ('Z') so a trailing \n can be
+                // flagged as CHAR_MUST_BREAK_AFTER, so we can show an empty line
+                // and allow the cursor to be positionned after that last \n.
+                int ch = i < m_length ? m_text[i] : 'Z';
+                int brk = lb_process_next_char(&lbCtx, ch);
+                // This tells us about a break between previous char and this 'ch'.
+                // printf("between <%c%c>: brk %d\n", m_text[i-1], m_text[i], brk);
+                // Note: LINEBREAK_ALLOWBREAK is set on the last space in a sequence
+                // of multiple consecutive spaces.
+                if ( brk == LINEBREAK_ALLOWBREAK ) {
+                    // Happens between a space (at i-1) and its following non-space
+                    // char, or after each CJK char.
+                    m_charinfo[i-1].flags |= CHAR_CAN_WRAP_AFTER;
+                    if ( m_text[i-1] == ' ' ) { // todo: also other kind of spaces?
```

Comment:
utf8proc *might* have a table of valid space codepoints.

Or something does. That kind of stuff sounds familiar, but I can't quite place it RN ;).

Reply:
Yes, and HB too, see https://github.com/koreader/crengine/blob/cb3b2da13b70d8bea86ef70bf959703b02f58962/crengine/src/lvtextfm.cpp#L1038-L1051.
But as there are [not many kind of space](http://jkorpela.fi/chars/spaces.html), we might as well just hardcode them here

Reply:
Ok, adressed that one with hardcoded char codes, and added another flag CAN_EXTEND_WIDTH_FALLBACK to mark CJK chars (actually, all allow_break chars), to allow text justification when there is not a single space.